contracts/modules/common/TokenPriceProvider.sol

```diff
@@ -16,33 +16,41 @@
 // SPDX-License-Identifier: GPL-3.0-only
 pragma solidity ^0.6.10;
 import "@openzeppelin/contracts/math/SafeMath.sol";
-import "../../../lib/other/ERC20.sol";
+import "../../../lib/other/IERC20Extended.sol";
 import "../../infrastructure/base/Managed.sol";
 
 contract TokenPriceProvider is Managed {
     using SafeMath for uint256;
 
     mapping(address => uint256) public cachedPrices;
 
-    function setPrice(ERC20 _token, uint256 _price) public onlyManager {
-        cachedPrices[address(_token)] = _price;
-    }
-
-    function setPriceForTokenList(ERC20[] calldata _tokens, uint256[] calldata _prices) external onlyManager {
-        for (uint16 i = 0; i < _tokens.length; i++) {
-            setPrice(_tokens[i], _prices[i]);
-        }
-    }
-
     /**
      * @dev Converts the value of _amount tokens in ether.
      * @param _amount the amount of tokens to convert (in 'token wei' twei)
-     * @param _token the ERC20 token contract
+     * @param _token the token address
      * @return the ether value (in wei) of _amount tokens with contract _token
      */
     function getEtherValue(uint256 _amount, address _token) external view returns (uint256) {
-        uint256 decimals = ERC20(_token).decimals();
+        uint8 decimals;
+
+        try IERC20Extended(_token).decimals() returns (uint8 _decimals) {
+            decimals = _decimals;
+        }
+        catch (bytes memory /*lowLevelData*/) {
+        }
+
         uint256 price = cachedPrices[_token];
-        return price.mul(_amount).div(10**decimals);
+        uint256 etherValue = decimals == 0 ? price.mul(_amount) : price.mul(_amount).div(10**uint(decimals));
```

Comment:
I'm a bit concerned about assuming `decimals == 0` when the `decimals` method is not present.

Looking for example at the DGD token (https://etherscan.io/token/0xE0B7927c4aF23765Cb51314A0E0521A9645F0E2A) I see that etherscan has the decimals set to 9 and this is the value we use in our token price provider. 

This means that calling `getEtherValue` will return a value 10**9 time bigger then what it should be. It is ok in terms of security (i.e. we are 10**9 times more drastic on the daily limit) but it is not consistent.

Maybe we could have a decimals mapping in our price provider for the few contracts that don't have the decimals method and use that value when the call to decimals fails?

Reply:
To be sure I wrote a script to pull all tokens listed in the provider with no `decimals` property and there's only the DGD one. The value of 9 for its decimals we see in etherscan is really just a free text they've added and it's probably dangerous to pretend that token contract has 9 decimals when it doesn't because the property can be used elsewhere in our contracts. I think the price in the provider should be set to actual price per token: 192297647 as opposed to adding 9 zeros to this which is what we appear to have done.

Reply:
Taking a step back, it would be simpler and cleaner not to have to deal with decimals at all. The reason we are using decimals in the TPP is because, in a previous version of the TPP, we fetched the rates from Kyber which, unfortunately, returned a price per "whole token" instead of returning a price per "token wei". Now that we no longer use Kyber, we could simply have `cachedPrices[_token]` be the rate such that `cachedPrices[_token].div(10**18)` is the number of weis per "token wei". In this case `getEtherValue()` would simply become:
```
function getEtherValue(uint256 _amount, address _token) external view returns (uint256) {
        uint256 price = cachedPrices[_token];
        return price.mul(_amount).div(10**18);
    }
``` 
This would require some coordination with backend and clients though. But maybe worth investing this time to simplify our lives in the long term.

Reply:
The problem is that it reduces accuracy for price per token wei and for 253 (out of 381) tokens that price even becomes zero.

Reply:
Copying here the examples discussed with Elena here for future reference:

Token= 0x595832f8fc6bf59c85c527fec3740a1b7a361269.
1 POWR = 0.000389118500000000 ETH.
Decimals = 6.
Currently we store cachedPrice = 389118500000000 = 0.000389118500000000*10^18  for that token.
With my proposal we would store cachedPrice = 0.000389118500000000 * 10^(36-6) = 389118500000000000000000000. And the getEtherValue would give:
 `getEtherValue(amount) = amount * cachedPrice / 10^18` (= 389118500 * amount)
whereas now it is calculated as:
`getEtherValue(amount) = amount * (0.000389118500000000*10^18) / 10^6` (= 389118500 * amount), which is equivalent.

Token= 0x05f4a42e251f2d52b8ed15e9fedaacfcef1fad27.
1 ZIL = 0.000079764579653685 ETH.
Decimals = 12.
Currently we store cachedPrice = 79764579653685 = 0.000079764579653685*10^18  for that token.
With my proposal we would store cachedPrice = 0.000079764579653685 * 10^(36-12) = 79764579653685000000. And the getEtherValue would give:
 `getEtherValue(amount) = amount * cachedPrice / 10^18` (~= 79.764579653685 * amount)
whereas now it is calculated as:
`getEtherValue(amount) = amount * (0.000079764579653685*10^18) / 10^12` (~= 79.764579653685 * amount), which is equivalent.

So essentially, the cachedPrices would just become whatever they are now multiplied by 10^(18-decimals) and getEtherValue would become `getEtherValue(amount) = amount * cachedPrice / 10^18` to take this change into account.